nimpy.nim

```diff
@@ -500,6 +518,91 @@ proc nimArrToPy[T](s: openarray[T]): PPyObject =
         let o = nimValueToPy(s[i])
         discard pyLib.PyList_SetItem(result, i, o)
 
+proc baseType(o: PPyObject): PyBaseType =
+    # returns the correct PyBaseType of the given PyObject extracted
+    # by manually checking all types
+    # If no call to `returnIfSubclass` returns from this proc, the
+    # default value of `pbUnknown` will be returned
+    template returnIfSubclass(pyt, nimt: untyped): untyped =
+        if checkObjSubclass(o, pyt):
+            return nimt
+
+    # check int types first for backward compatibility with Python2
+    returnIfSubclass(Py_TPFLAGS_INT_SUBCLASS or Py_TPFLAGS_LONG_SUBCLASS, pbLong)
+
+    let checkTypes = { pyLib.PyFloat_Type : pbFloat,
+                       pyLib.PyComplex_Type : pbComplex,
+                       pyLib.PyBytes_Type : pbString,
+                       pyLib.PyUnicode_Type : pbString,
+                       pyLib.PyList_Type : pbList,
+                       pyLib.PyTuple_Type : pbTuple,
+                       pyLib.PyDict_Type : pbDict }
+
+    for tup in checkTypes:
+        let
+            k = tup[0]
+            v = tup[1]
+        returnIfSubclass(k, v)
+    # if we have not returned until here, `pbUnknown` is returned
+
+iterator items*(o: PyObject): PyObject =
+    let it = pyLib.PyObject_GetIter(o.rawPyObj)
+    while true:
+        let i = pyLib.PyIter_Next(it)
+        if i.isNil: break
+        yield newPyObjectConsumingRef(i)
+    decRef it
+
+proc `$`*(o: PyObject): string
+proc `$`*(p: PPyObject): string =
```

Comment:
Use `pyObjToNimStr` to avoid extra allocation, and memory leak which you've got here.

Reply:
The extra allocation I understand, but could you explain why this causes a memory leak? 
In the scope of the `$` proc for `PPyObject` I only create a `PyObject`, which should be GC'ed, since it's only a Nim object, no? Sorry, I'm still pretty unclear about Python's API.

Reply:
The extra allocation I understand, but could you explain why this causes a memory leak? 
In the scope of the `$` proc for `PPyObject` I only create a `PyObject`, which should be GC'ed, since it's only a Nim object, no? Sorry, I'm still pretty unclear about Python's API.

Reply:
My bad, you're right. Nothing bad, just an extra allocation.